src/cpp/ext/gsm/metadata_exchange.cc

```diff
@@ -120,6 +122,116 @@ absl::string_view GetStringValueFromUpbStruct(google_protobuf_Struct* struct_pb,
   return "unknown";
 }
 
+class LocalLabelsIterable : public LabelsIterable {
+ public:
+  explicit LocalLabelsIterable(
+      const std::vector<std::pair<absl::string_view, std::string>>& labels)
+      : labels_(labels) {}
+
+  absl::optional<std::pair<absl::string_view, absl::string_view>> Next()
+      override {
+    if (pos_ >= labels_.size()) {
+      return absl::nullopt;
+    }
+    return labels_[pos_++];
+  }
+
+  size_t Size() const override { return labels_.size(); }
+
+  void ResetIteratorPosition() override { pos_ = 0; }
+
+ private:
+  size_t pos_ = 0;
+  const std::vector<std::pair<absl::string_view, std::string>>& labels_;
+};
+
+class PeerLabelsIterable : public LabelsIterable {
+ public:
+  explicit PeerLabelsIterable(grpc_core::Slice remote_metadata) {
+    std::string decoded_metadata;
```

Comment:
Instead of doing this decoding here, I suggest doing it in `Next()` when `pos_` is 0.  That way, we're not holding on to the memory for the deserialized proto until we actually need it.  Note that for a streaming call, there can be a long time between receiving the initial metadata and the end of the call, which is where we actually need to do the iteration.

Reply:
We either hold on to the metadata or the deserialized proto. 
Note that for `Size()`, we need the deserialized proto anyway. We *could* do it lazily the first time we need it?

Reply:
Changed to do it lazily